In [60]:
import pandas as pd

In [61]:
df = pd.read_csv('McDonalds.csv') # read the csv file
df

Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving Size  Calories  Calories from Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total Fat (% Daily Value)  Saturated Fat  Saturated Fat (% Daily Value)  \
0                           20            5.0                             25   
1                           12            3.0                             15   
2                           35            8.0                             42   
3                           43           10.0                             52   
4                           35            8.0                             42   
..                         ...            ...                            ...   
255                         26            9.0                             44   
256                         35           12.0                             58   
257                         17            6.0                             29   
258                         50           15.0                             76   
259                         25            8.0                             38   

     Trans Fat  ...  Carbohydrates  Carbohydrates (% Daily Value)  \
0          0.0  ...             31                             10   
1          0.0  ...             30                             10   
2          0.0  ...             29                             10   
3          0.0  ...             30                             10   
4          0.0  ...             30                             10   
..         ...  ...            ...                            ...   
255        0.5  ...             80                             27   
256        1.0  ...            106                             35   
257        0.0  ...             53                             18   
258        1.0  ...            114                             38   
259        0.0  ...             57                             19   

     Dietary Fiber  Dietary Fiber (% Daily Value)  Sugars  Protein  \
0                4                             17       3       17   
1                4                             17       3       18   
2                4                             17       2       14   
3                4                             17       2       21   
4                4                             17       2       21   
..             ...                            ...     ...      ...   
255              1                       

In [62]:
df.shape # check the shape of the data

(260, 24)

In [63]:
df.size # check the size of the data

6240

In [64]:
df.columns


Index(['Category', 'Item', 'Serving Size', 'Calories', 'Calories from Fat',
       'TotalFat', 'Total Fat (% Daily Value)', 'Saturated Fat',
       'Saturated Fat (% Daily Value)', 'Trans Fat', 'Cholesterol',
       'Cholesterol (% Daily Value)', 'Sodium', 'Sodium (% Daily Value)',
       'Carbohydrates', 'Carbohydrates (% Daily Value)', 'Dietary Fiber',
       'Dietary Fiber (% Daily Value)', 'Sugars', 'Protein',
       'Vitamin A (% Daily Value)', 'Vitamin C (% Daily Value)',
       'Calcium (% Daily Value)', 'Iron (% Daily Value)'],
      dtype='object')

In [65]:
df.dtypes


Category                          object
Item                              object
Serving Size                      object
Calories                           int64
Calories from Fat                  int64
TotalFat                         float64
Total Fat (% Daily Value)          int64
Saturated Fat                    float64
Saturated Fat (% Daily Value)      int64
Trans Fat                        float64
Cholesterol                        int64
Cholesterol (% Daily Value)        int64
Sodium                             int64
Sodium (% Daily Value)             int64
Carbohydrates                      int64
Carbohydrates (% Daily Value)      int64
Dietary Fiber                      int64
Dietary Fiber (% Daily Value)      int64
Sugars                             int64
Protein                            int64
Vitamin A (% Daily Value)          int64
Vitamin C (% Daily Value)          int64
Calcium (% Daily Value)            int64
Iron (% Daily Value)               int64
dtype: object

## How we can change column names?
What kind of problems we can see with this data set?

- we have spaces in column names what will cause problems if we'd like to access the using object notation.



In [66]:
#df.columns[0] = 'cat' # KEY ERROR: 'Index' object does not support item assignment 
# because the columns are immutable
# TypeError("Index does not support mutable operations")

Instead we could've try to assign a new list with all the new column names. Although it will work it's cumbersome approach, beucase we need to list all 24 columns... Instead we can use a trick with list comprehensions.



In [67]:
# what is the cumnersome approach to change the column name?
# 1. create a list of column names
# 2. change the name of the first element in the list
# 3. assign the list to the columns attribute of the dataframe

df.columns = [column_name.upper() for column_name in df.columns]
# the original column names are in lower case # dataframes get modified
df

CATEGORY                                               ITEM  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        SERVING SIZE  CALORIES  CALORIES FROM FAT  TOTALFAT  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     TOTAL FAT (% DAILY VALUE)  SATURATED FAT  SATURATED FAT (% DAILY VALUE)  \
0                           20            5.0                             25   
1                           12            3.0                             15   
2                           35            8.0                             42   
3                           43           10.0                             52   
4                           35            8.0                             42   
..                         ...            ...                            ...   
255                         26            9.0                             44   
256                         35           12.0                             58   
257                         17            6.0                             29   
258                         50           15.0                             76   
259                         25            8.0                             38   

     TRANS FAT  ...  CARBOHYDRATES  CARBOHYDRATES (% DAILY VALUE)  \
0          0.0  ...             31                             10   
1          0.0  ...             30                             10   
2          0.0  ...             29                             10   
3          0.0  ...             30                             10   
4          0.0  ...             30                             10   
..         ...  ...            ...                            ...   
255        0.5  ...             80                             27   
256        1.0  ...            106                             35   
257        0.0  ...             53                             18   
258        1.0  ...            114                             38   
259        0.0  ...             57                             19   

     DIETARY FIBER  DIETARY FIBER (% DAILY VALUE)  SUGARS  PROTEIN  \
0                4                             17       3       17   
1                4                             17       3       18   
2                4                             17       2       14   
3                4                             17       2       21   
4                4                             17       2       21   
..             ...                            ...     ...      ...   
255              1                       

In [68]:
df = pd.read_csv('https://raw.githubusercontent.com/MuhammetBilgic/Numpy_Excercises/main/ALX_day_7/McDonalds.csv')
df

Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving Size  Calories  Calories from Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total Fat (% Daily Value)  Saturated Fat  Saturated Fat (% Daily Value)  \
0                           20            5.0                             25   
1                           12            3.0                             15   
2                           35            8.0                             42   
3                           43           10.0                             52   
4                           35            8.0                             42   
..                         ...            ...                            ...   
255                         26            9.0                             44   
256                         35           12.0                             58   
257                         17            6.0                             29   
258                         50           15.0                             76   
259                         25            8.0                             38   

     Trans Fat  ...  Carbohydrates  Carbohydrates (% Daily Value)  \
0          0.0  ...             31                             10   
1          0.0  ...             30                             10   
2          0.0  ...             29                             10   
3          0.0  ...             30                             10   
4          0.0  ...             30                             10   
..         ...  ...            ...                            ...   
255        0.5  ...             80                             27   
256        1.0  ...            106                             35   
257        0.0  ...             53                             18   
258        1.0  ...            114                             38   
259        0.0  ...             57                             19   

     Dietary Fiber  Dietary Fiber (% Daily Value)  Sugars  Protein  \
0                4                             17       3       17   
1                4                             17       3       18   
2                4                             17       2       14   
3                4                             17       2       21   
4                4                             17       2       21   
..             ...                            ...     ...      ...   
255              1                       

I'd like to do several changes to column names:

- replace spaces by `_`
- replace `%` with 'Percent'
- replace `(` with empty string
- replace `)` with empty string 

In [69]:
my_str = 'Total Fat (% Daily Value)'
my_str.replace(' ', '_').replace('(', '') .replace('%', 'Percent') # method chaining

'Total_Fat_Percent_Daily_Value)'

In [76]:
df.columns = [column_name.replace(' ', '_')
              .replace('(', '')
              .replace(')', '')
              .replace('%', 'Percent') for column_name in df.columns] # method chaining 
df

Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12            3.0   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1   

[`.rename()`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html) -  we can use rename to change column names and this is another operation with **inplace** argument, so by default returns a copy and doesn't modify the original data.

In [77]:
df.rename(columns={'Category': 'Item_Category'}) # rename a single column

Item_Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12            3.0   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   

In [79]:
df # the original dataframe is not modified

Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12            3.0   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1   

In [80]:
df.rename(columns={'Category': 'Product_Category', 'Item': 'Product'})


Product_Category                                            Product  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12            3.0   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17

## Data modifications

We have several dedicated methods to change the data within DataFrame or Series objects:

- [**.replace()**](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html) - will allow to change specified value into a different one
- [**fillna()**](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html) - used to fill/replace [missing values](https://pandas.pydata.org/docs/user_guide/missing_data.html)
- **.apply()** - allows us to apply any function (either Python, Pandas or our own function) to the data


### replace - "find and replace"

We can apply it to the whole DataFrame of to the particular Series. This function also returns copy by default, which we can change via **inplace** argument.



In [81]:
df.replace('Breakfast', 'Petit-dejeuner')


Category                                               Item  \
0        Petit-dejeuner                                       Egg McMuffin   
1        Petit-dejeuner                                  Egg White Delight   
2        Petit-dejeuner                                   Sausage McMuffin   
3        Petit-dejeuner                          Sausage McMuffin with Egg   
4        Petit-dejeuner                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12            3.0   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1   

In [82]:
df.replace(23.0, -99)
#           float , int


Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200     -99.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210     -99.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200     -99.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12            3.0   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1   

In [83]:
df.Saturated_Fat.replace(3.0, -10.5, inplace=True)
df # the original dataframe is modified based on the inplace argument

Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12          -10.5   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1   

In [85]:
df.Category.unique() # unique values in a column


array(['Breakfast', 'Beef & Pork', 'Chicken & Fish', 'Salads',
       'Snacks & Sides', 'Desserts', 'Beverages', 'Coffee & Tea',
       'Smoothies & Shakes'], dtype=object)

In [86]:
df.replace(['Desserts', 'Smoothies & Shakes'], 'Sweets') # replace multiple values to a single value


Category                                               Item  \
0    Breakfast                                       Egg McMuffin   
1    Breakfast                                  Egg White Delight   
2    Breakfast                                   Sausage McMuffin   
3    Breakfast                          Sausage McMuffin with Egg   
4    Breakfast                   Sausage McMuffin with Egg Whites   
..         ...                                                ...   
255     Sweets                 McFlurry with Oreo Cookies (Small)   
256     Sweets                McFlurry with Oreo Cookies (Medium)   
257     Sweets                 McFlurry with Oreo Cookies (Snack)   
258     Sweets  McFlurry with Reese's Peanut Butter Cups (Medium)   
259     Sweets   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12          -10.5   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1                                   17       3       18   
2                                   17      

In [87]:
df.Category.unique()


array(['Breakfast', 'Beef & Pork', 'Chicken & Fish', 'Salads',
       'Snacks & Sides', 'Desserts', 'Beverages', 'Coffee & Tea',
       'Smoothies & Shakes'], dtype=object)

In [88]:
df.replace(['Breakfast', 'Beverages'], ['Petit-dejeuner', 'Beuverages']) # replace multiple values to multiple values


Category                                               Item  \
0        Petit-dejeuner                                       Egg McMuffin   
1        Petit-dejeuner                                  Egg White Delight   
2        Petit-dejeuner                                   Sausage McMuffin   
3        Petit-dejeuner                          Sausage McMuffin with Egg   
4        Petit-dejeuner                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12          -10.5   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1   

In [89]:
df.replace({
    'Category': ['Breakfast', 'Beverages'],
    'Item': ['Egg McMuffin']
}, 'ABCD') # replace multiple values to multiple values

Category                                               Item  \
0                  ABCD                                               ABCD   
1                  ABCD                                  Egg White Delight   
2                  ABCD                                   Sausage McMuffin   
3                  ABCD                          Sausage McMuffin with Egg   
4                  ABCD                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12          -10.5   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1   

In [90]:
df.replace({
    'Category': ['Breakfast', 'Beverages'],
    'Item': ['Egg McMuffin']
}, {
    'Category': 'AAA',
    'Item': 'BBB'
})
# replace multiple values to multiple values 

Category                                               Item  \
0                   AAA                                                BBB   
1                   AAA                                  Egg White Delight   
2                   AAA                                   Sausage McMuffin   
3                   AAA                          Sausage McMuffin with Egg   
4                   AAA                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12          -10.5   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1   

We can use regular expression within .**replace()**. We have to enable them by adding **regex=True** named argument.

By default **replace** is finding and replacing whole values, not fragments of them. To make that happen we can enable regular expression and then we will be able to change fragments of texts.




In [94]:
df.replace('Egg', 'Oeuf', regex=True)


Category                                               Item  \
0             Breakfast                                      Oeuf McMuffin   
1             Breakfast                                 Oeuf White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                         Sausage McMuffin with Oeuf   
4             Breakfast                  Sausage McMuffin with Oeuf Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12          -10.5   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1   

If we are working with regular expression in Python code, especially providing patterns, then it's best to use r-string or raw strings: r''. This means that Python will not try to change any strings inside of the string (/).



In [96]:
df.replace(r'Mc(\w+)', r'\1', regex=True)


Category                                             Item  \
0             Breakfast                                       Egg Muffin   
1             Breakfast                                Egg White Delight   
2             Breakfast                                   Sausage Muffin   
3             Breakfast                          Sausage Muffin with Egg   
4             Breakfast                   Sausage Muffin with Egg Whites   
..                  ...                                              ...   
255  Smoothies & Shakes                 Flurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                Flurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 Flurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  Flurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   Flurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12          -10.5   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Carbohydrates  \
0                                   25        0.0  ...             31   
1                                   15        0.0  ...             30   
2                                   42        0.0  ...             29   
3                                   52        0.0  ...             30   
4                                   42        0.0  ...             30   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...             80   
256                                 58        1.0  ...            106   
257                                 29        0.0  ...             53   
258                                 76        1.0  ...            114   
259                                 38        0.0  ...             57   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1                           

## **apply**

Applies a function to each value from a DataFrame (row) or Series.

Calories                 New series
300  -> times_10(300) -> 3000
250  -> times_10(250) -> 2500
370  -> times_10(370) -> 3700

In [97]:
def times_10(number):
    return number * 10

In [98]:
times_10(5)


50

When we want to say that apply should use a particular function (either our function defined by **def** keyword or any built-in function) we just provide the name of the function **apply(times_10)** and I'm NOT adding **()** after the name of the function **apply(times_10())** (incorrect), becuase adding **()** means that I'm executing, I'm running the function.



In [100]:
df.Calories.apply(times_10) # apply a function to a column


0      3000
1      2500
2      3700
3      4500
4      4000
       ... 
255    5100
256    6900
257    3400
258    8100
259    4100
Name: Calories, Length: 260, dtype: int64

In [101]:
df.Calories.apply(lambda number: number *10)

0      3000
1      2500
2      3700
3      4500
4      4000
       ... 
255    5100
256    6900
257    3400
258    8100
259    4100
Name: Calories, Length: 260, dtype: int64

How we can extract information about grams from a *Serving_Size* column?


In [106]:
import re # regular expression

def extract_grams(serving_size: str) -> float:
    matches = re.search(r'\((\d+) g\)', serving_size)
    if matches:
        return float(matches[1])
    else:
        return 0.0
    
df.Serving_Size, df.Serving_Size.apply(extract_grams)



(0       4.8 oz (136 g)
 1       4.8 oz (135 g)
 2       3.9 oz (111 g)
 3       5.7 oz (161 g)
 4       5.7 oz (161 g)
             ...       
 255    10.1 oz (285 g)
 256    13.4 oz (381 g)
 257     6.7 oz (190 g)
 258    14.2 oz (403 g)
 259     7.1 oz (202 g)
 Name: Serving_Size, Length: 260, dtype: object,
 0      136.0
 1      135.0
 2      111.0
 3      161.0
 4      161.0
        ...  
 255    285.0
 256    381.0
 257    190.0
 258    403.0
 259    202.0
 Name: Serving_Size, Length: 260, dtype: float64)

In [109]:
df['Serving_Size_Grams'] = df.Serving_Size.apply(extract_grams) # create a new column based on the result of the apply method
df

Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12          -10.5   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  \
0                                   25        0.0  ...   
1                                   15        0.0  ...   
2                                   42        0.0  ...   
3                                   52        0.0  ...   
4                                   42        0.0  ...   
..                                 ...        ...  ...   
255                                 44        0.5  ...   
256                                 58        1.0  ...   
257                                 29        0.0  ...   
258                                 76        1.0  ...   
259                                 38        0.0  ...   

     Carbohydrates_Percent_Daily_Value  Dietary_Fiber  \
0                                   10              4   
1                                   10              4   
2                                   10              4   
3                                   10              4   
4                                   10              4   
..                                 ...            ...   
255                                 27              1   
256                                 35              1   
257                                 18              1   
258                                 38              2   
259                                 19              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1                                   17       3       18   
2                                   17       2       14   
3                                   17       2       21   
4      

In [116]:
df['Saturation'] = df['Calories'] / df['Serving_Size_Grams'] # create a new column based on the result of the apply method
df

Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
..                  ...                                                ...   
255  Smoothies & Shakes                 McFlurry with Oreo Cookies (Small)   
256  Smoothies & Shakes                McFlurry with Oreo Cookies (Medium)   
257  Smoothies & Shakes                 McFlurry with Oreo Cookies (Snack)   
258  Smoothies & Shakes  McFlurry with Reese's Peanut Butter Cups (Medium)   
259  Smoothies & Shakes   McFlurry with Reese's Peanut Butter Cups (Snack)   

        Serving_Size  Calories  Calories_from_Fat  TotalFat  \
0     4.8 oz (136 g)       300                120      13.0   
1     4.8 oz (135 g)       250                 70       8.0   
2     3.9 oz (111 g)       370                200      23.0   
3     5.7 oz (161 g)       450                250      28.0   
4     5.7 oz (161 g)       400                210      23.0   
..               ...       ...                ...       ...   
255  10.1 oz (285 g)       510                150      17.0   
256  13.4 oz (381 g)       690                200      23.0   
257   6.7 oz (190 g)       340                100      11.0   
258  14.2 oz (403 g)       810                290      32.0   
259   7.1 oz (202 g)       410                150      16.0   

     Total_Fat_Percent_Daily_Value  Saturated_Fat  \
0                               20            5.0   
1                               12          -10.5   
2                               35            8.0   
3                               43           10.0   
4                               35            8.0   
..                             ...            ...   
255                             26            9.0   
256                             35           12.0   
257                             17            6.0   
258                             50           15.0   
259                             25            8.0   

     Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  Dietary_Fiber  \
0                                   25        0.0  ...              4   
1                                   15        0.0  ...              4   
2                                   42        0.0  ...              4   
3                                   52        0.0  ...              4   
4                                   42        0.0  ...              4   
..                                 ...        ...  ...            ...   
255                                 44        0.5  ...              1   
256                                 58        1.0  ...              1   
257                                 29        0.0  ...              1   
258                                 76        1.0  ...              2   
259                                 38        0.0  ...              1   

     Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
0                                   17       3       17   
1                                   17       3       18   
2                                   17       2       14   
3                                   17       2       21   
4                                   17       2       21   
..                                 ...     ...      ...   
255                                  4      64       12   
256                                  5      85       15   
257                                  2      43        8   
258                                  9     103       21   
259                                  5      51       10   

     Vitamin_A_Percent_Daily_Value  Vitamin_C_Percent_Daily_Value  \
0                            

In [112]:
df.sort_values('Saturation', ascending=False)


Category                          Item       Serving_Size  Calories  \
130     Beverages           1% Low Fat Milk Jug  1 carton (236 ml)       100   
157  Coffee & Tea  French Vanilla Latte (Small)       12 fl oz cup       260   
144  Coffee & Tea             Sweet Tea (Child)       12 fl oz cup       110   
148  Coffee & Tea                 Latte (Small)       12 fl oz cup       170   
149  Coffee & Tea                Latte (Medium)       16 fl oz cup       210   
..            ...                           ...                ...       ...   
139  Coffee & Tea              Iced Tea (Large)       30 fl oz cup         0   
140  Coffee & Tea              Iced Tea (Child)       12 fl oz cup         0   
145  Coffee & Tea                Coffee (Small)       12 fl oz cup         0   
146  Coffee & Tea               Coffee (Medium)       16 fl oz cup         0   
147  Coffee & Tea                Coffee (Large)       16 fl oz cup         0   

     Calories_from_Fat  TotalFat  Total_Fat_Percent_Daily_Value  \
130                 20       2.5                              4   
157                 80       9.0                             13   
144                  0       0.0                              0   
148                 80       9.0                             13   
149                 90      10.0                             16   
..                 ...       ...                            ...   
139                  0       0.0                              0   
140                  0       0.0                              0   
145                  0       0.0                              0   
146                  0       0.0                              0   
147                  0       0.0                              0   

     Saturated_Fat  Saturated_Fat_Percent_Daily_Value  Trans_Fat  ...  \
130            1.5                                  8        0.0  ...   
157            5.0                                 24        0.0  ...   
144            0.0                                  0        0.0  ...   
148            5.0                                 24        0.0  ...   
149            6.0                                 30        0.0  ...   
..             ...                                ...        ...  ...   
139            0.0                                  0        0.0  ...   
140            0.0                                  0        0.0  ...   
145            0.0                                  0        0.0  ...   
146            0.0                                  0        0.0  ...   
147            0.0                                  0        0.0  ...   

     Dietary_Fiber  Dietary_Fiber_Percent_Daily_Value  Sugars  Protein  \
130              0                                  0      12        8   
157              1                                  3      36        9   
144              0                                  0      27        0   
148              1                                  3      12        9   
149              1                                  4      15       11   
..             ...                                ...     ...      ...   
139              0                                  0       0        0   
140              0                                  0       0        0   
145              0                                  0       0        0   
146              0                                  0       0        0   
147              0                                  0       0        0   

     Vitamin_A_Percent_Daily_Value  Vitamin_C_Percent_Daily_Value  \
130                             10                              4   
157                              8                              0   
144                              0                              0   
148                              8                              0   
149                             10                              0   
..                             ...                            .